File: src/main/java/ml/duncte123/skybot/commands/uncategorized/UserinfoCommand.java
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan & Maurice R S "Sanduhr32"
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.uncategorized;

import ml.duncte123.skybot.objects.command.Command;
import ml.duncte123.skybot.utils.AirUtils;
import ml.duncte123.skybot.utils.EmbedUtils;
import ml.duncte123.skybot.utils.Settings;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.MessageEmbed;
import net.dv8tion.jda.core.entities.User;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import org.apache.commons.lang3.StringUtils;

import java.time.format.DateTimeFormatter;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by Duncan on 9-7-2017.
 */
public class UserinfoCommand extends Command {
    
    @Override
    public void executeCommand(String invoke, String[] args, GuildMessageReceivedEvent event) {
        User u;
        Member m;
        
        if (args.length == 0) {
            u = event.getAuthor();
            m = event.getGuild().getMemberById(u.getId());
        } else {
            List<User> mentioned = event.getMessage().getMentionedUsers();
            if (!mentioned.isEmpty()) {
                m = event.getGuild().getMember(mentioned.get(0));
            } else {
                String name = StringUtils.join(args, " ");
                
                List<Member> members = event.getGuild().getMembersByName(name, true);
                if (members.isEmpty()) {
                    members = event.getGuild().getMembersByNickname(name, true);
                }
                m = members.isEmpty() ? null : members.get(0);
            }
        }
        
        if (m == null) {
            event.getChannel().sendMessage("This user could not be found.").queue();
            return;
        }
        
        u = m.getUser();
        
        StringBuilder joinOrder = new StringBuilder();
        List<Member> joins = event.getGuild().getMemberCache().parallelStream().collect(Collectors.toList());
        joins.sort(Comparator.comparing(Member::getJoinDate));
        int index = joins.indexOf(m);
        index -= 3;
        if (index < 0)
            index = 0;
        joinOrder.append("\n");
        if (joins.get(index).equals(m))
            joinOrder.append("[").append(joins.get(index).getEffectiveName()).append("](https://bot.duncte123.me/)");
        else
            joinOrder.append(joins.get(index).getEffectiveName());
        for (int i = index + 1; i < index + 7; i++) {
            if (i >= joins.size())
                break;
            Member usr = joins.get(i);
            String name = usr.getEffectiveName();
            if (usr.equals(m))
                name = "[" + name + "](https://bot.duncte123.me/)";
            joinOrder.append(" > ").append(name);
        }
        
        //TODO: make request to discord profiles
        
        MessageEmbed eb = EmbedUtils.defaultEmbed()
                                  .setColor(m.getColor())
                                  .setDescription("User info for " + u.getName() + "#" + u.getDiscriminator())
                                  .setThumbnail(u.getEffectiveAvatarUrl())
                                  .addField("Username + Discriminator", String.format("%#s", u), true)
                                  .addField("User Id", u.getId(), true)
                                  .addField("Status", AirUtils.gameToString(m.getGame()), true)
                                  .addField("Nickname", (m.getNickname() == null ? "**_no nickname_**" : m.getNickname()), true)
                                  .addField("Created", u.getCreationTime().format(DateTimeFormatter.RFC_1123_DATE_TIME), true)
                                  .addField("Joined", m.getJoinDate().format(DateTimeFormatter.RFC_1123_DATE_TIME), true)
                                  .addField("Join order", joinOrder.toString(), true)
                                  .addField("Online Status", AirUtils.convertStatus(m.getOnlineStatus()) + " " + m.getOnlineStatus().name().toLowerCase().replaceAll("_", " "), true)
                                  .addField("Is a bot", (u.isBot() ? "Yep, this user is a bot" : "Nope, this user is not a bot"), true)
                                  .build();
        
        sendEmbed(event, eb);
    }
    
    @Override
    public String help() {
        return "Get information from yourself or from another user.\nUsage: `" + Settings.prefix + getName() + " [username]`";
    }
    
    @Override
    public String getName() {
        return "userinfo";
    }
    
    @Override
    public String[] getAliases() {
        return new String[]{"user", "i"};
    }
}

File: src/main/kotlin/ml/duncte123/skybot/commands/uncategorized/UserinfoCommand.kt
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan & Maurice R S "Sanduhr32"
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.uncategorized

import me.duncte123.botcommons.messaging.EmbedUtils
import me.duncte123.botcommons.messaging.MessageUtils.sendEmbed
import me.duncte123.botcommons.messaging.MessageUtils.sendMsg
import me.duncte123.weebJava.types.StatusType
import ml.duncte123.skybot.Settings.PATREON
import ml.duncte123.skybot.entities.jda.DunctebotGuild
import ml.duncte123.skybot.extensions.*
import ml.duncte123.skybot.objects.CooldownScope
import ml.duncte123.skybot.objects.Emotes.*
import ml.duncte123.skybot.objects.command.Command
import ml.duncte123.skybot.objects.command.CommandContext
import ml.duncte123.skybot.utils.FinderUtils
import net.dv8tion.jda.api.OnlineStatus
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.User.UserFlag
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import java.time.OffsetDateTime
import java.time.temporal.ChronoUnit
import kotlin.streams.toList

class UserinfoCommand : Command() {
    private val nitroUserLink = "**[Nitro User:](https://github.com/DuncteBot/SkyBot/issues/201#issuecomment-486182959 \"Click for more info on the nitro user check\")**"

    init {
        this.name = "userinfo"
        this.aliases = arrayOf("user", "i", "whois", "ui", "retrieveuserinfo")
        this.help = "Get some information about yourself or from another user"
        this.usage = "[@user]"
        this.cooldown = 30
        this.cooldownScope = CooldownScope.GUILD
    }

    override fun execute(ctx: CommandContext) {
        val event = ctx.event
        val args = ctx.args

        if (ctx.invoke.lowercase() == "retrieveuserinfo") {
            if (args.isEmpty()) {
                sendMsg(ctx, "Missing arguments for retrieving user information")

                return
            }

            ctx.jda.retrieveUserById(args[0]).queue(
                {
                    renderUserEmbed(ctx, it, ctx.guild)
                },
                {
                    sendMsg(ctx, "Could not get user info: ${it.message}")
                }
            )

            return
        }

        var u: User? = null
        var m: Member? = null

        if (args.isEmpty()) {
            u = event.author
            m = event.member
        } else {
            val members = FinderUtils.searchMembers(ctx.argsRaw, ctx)

            if (members.isNotEmpty()) {
                m = members[0]
                u = m.user
            } else {
                val users = FinderUtils.searchUsers(ctx.argsRaw, ctx)

                if (users.isNotEmpty()) {
                    u = users[0]
                }
            }
        }

        // if we have a user but not a member
        if (u != null && m == null) {
            renderUserEmbed(ctx, u, ctx.guild)
            return
        }

        if (m == null) {
            sendMsg(ctx, "This user could not be found.")
            return
        }

        renderMemberEmbed(event, m, ctx)
    }

    private fun renderUserEmbed(ctx: CommandContext, user: User, guild: DunctebotGuild) {
        val times = user.parseTimeCreated()

        val embed = EmbedUtils.getDefaultEmbed()
            .setColor(guild.color)
            .setThumbnail(user.getStaticAvatarUrl())
            .setDescription(
                """User info for ${user.asMention} ${user.badgeLine}
                        |
                        |**User Tag:** ${user.asTag.escapeMarkDown()}
                        |**User Id:** ${user.id}
                        |**Account Created:** ${times.first} (${times.second})
                        |$nitroUserLink ${user.isNitro.toEmoji()}
                        |**Bot Account:** ${user.isBot.toEmoji()}
                        |
                        |_Use `${guild.settings.customPrefix}avatar [user]` to get a user's avatar_
                    """.trimMargin()
            )

        sendEmbed(ctx, embed)
    }

    private fun generateJoinOrder(members: List<Member>, member: Member) = buildString {
        val joins = members.stream().sorted(Comparator.comparing(Member::getTimeJoined)).toList()

        var index = joins.indexOf(member)
        index -= 3

        if (index < 0) {
            index = 0
        }

        appendLine()

        if (joins[index] == member) {
            append("[${joins[index].effectiveName.escapeMarkDown()}]($PATREON)")
        } else {
            append(joins[index].effectiveName.escapeMarkDown())
        }

        for (i in index + 1 until index + 7) {
            if (i >= joins.size) {
                break
            }

            val mbr = joins[i]
            var usrName = mbr.effectiveName.escapeMarkDown()

            if (mbr == member) {
                usrName = "[$usrName]($PATREON)"
            }

            append(" \\> ")
            append(usrName)
        }
    }

    private fun getJoinPosition(members: List<Member>, member: Member): Long {
        return members.stream().sorted(Comparator.comparing(Member::getTimeJoined))
            .takeWhile {
                it != member
            }.count() + 1
    }

    private fun renderMemberEmbed(event: GuildMessageReceivedEvent, member: Member, ctx: CommandContext) {
        val user = member.user

        val userTimes = user.parseTimeCreated()
        val memberTimes = member.parseTimeJoined()
        var boostEmote = ""

        val boostingSinceMsg = if (member.timeBoosted == null) {
            ""
        } else {
            val boostTime = member.timeBoosted!!
            val boostTimes = boostTime.parseTimes()

            boostEmote = boostTime.toBoostEmote()

            "\n**Boosting since:** ${boostTimes.first} (${boostTimes.second})"
        }

        val userNitro = user.isNitro
        val nitroBadge = if (userNitro) " $DISCORD_NITRO" else ""
        val loadedMembers = event.guild.loadMembers().get()

        val embed = EmbedUtils.getDefaultEmbed()
            .setColor(member.color)
            .setThumbnail(user.getStaticAvatarUrl())
            .setDescription(
                """User info for ${member.asMention}$nitroBadge ${user.badgeLine} $boostEmote
                        |
                        |**User Tag:** ${user.asTag.escapeMarkDown()}
                        |**User Id:** ${user.id}
                        |**Display Name:** ${member.effectiveName.escapeMarkDown()}
                        |**Account Created:** ${userTimes.first} (${userTimes.second})
                        |$nitroUserLink ${userNitro.toEmoji()}
                        |**Joined Server:** ${memberTimes.first} (${memberTimes.second})
                        |**Join position:** #${getJoinPosition(loadedMembers, member)}
                        |**Join Order:** ${generateJoinOrder(loadedMembers, member)}
                        |**Bot Account:** ${user.isBot.toEmoji()}
                        |**Boosting:** ${(member.timeBoosted != null).toEmoji()}$boostingSinceMsg
                        |
                        |_Use `${ctx.prefix}avatar [user]` to get a user's avatar_
                    """.trimMargin()
            )

        // If we don't have permission to send files or our weebSh key is null
        if (!ctx.selfMember.hasPermission(event.channel, Permission.MESSAGE_ATTACH_FILES) ||
            ctx.config.apis.weebSh == null
        ) {
            sendEmbed(ctx, embed, true)
            return
        }

        ctx.weebApi.generateDiscordStatus(
            StatusType.values().random(),
            user.getStaticAvatarUrl() + "?size=256"
        ).async {
            event.channel.sendFile(it, "stat.png")
                .embed(embed.setThumbnail("attachment://stat.png").build())
                .queue(null) {
                    sendEmbed(ctx, embed.setThumbnail(user.effectiveAvatarUrl), true)
                }
        }
    }

    private fun toWeebshStatus(member: Member): StatusType {
        if (member.activities.isNotEmpty() && member.activities.any { it.type == Activity.ActivityType.STREAMING }) {
            return StatusType.STREAMING
        }

        return when (member.onlineStatus) {
            OnlineStatus.ONLINE -> StatusType.ONLINE
            OnlineStatus.DO_NOT_DISTURB -> StatusType.DND
            OnlineStatus.IDLE -> StatusType.IDLE
            else -> StatusType.OFFLINE
        }
    }

    private val User.isNitro: Boolean
        get() = this.avatarId != null && (this.avatarId as String).startsWith("a_")

    private val User.badgeLine: String
        get() = this.flags.mapNotNull { it.toEmote() }.joinToString(" ")

    private fun UserFlag.toEmote(): String? {
        return when (this) {
            UserFlag.STAFF -> DISCORD_STAFF
            UserFlag.PARTNER -> DISCORD_PARTNER

            UserFlag.HYPESQUAD -> DISCORD_HYPESQUAD
            UserFlag.HYPESQUAD_BRAVERY -> DISCORD_HYPESQUAD_BRAVERY
            UserFlag.HYPESQUAD_BRILLIANCE -> DISCORD_HYPESQUAD_BRILLIANCE
            UserFlag.HYPESQUAD_BALANCE -> DISCORD_HYPESQUAD_BALANCE

            UserFlag.BUG_HUNTER_LEVEL_1 -> DISCORD_BUG_HUNTER_1
            UserFlag.BUG_HUNTER_LEVEL_2 -> DISCORD_BUG_HUNTER_2

            UserFlag.EARLY_SUPPORTER -> DISCORD_EARLY_SUPPORTER
            // No emotes / not needed
//            UserFlag.TEAM_USER -> ""
//            UserFlag.SYSTEM -> ""
//            UserFlag.VERIFIED_BOT -> ""
            UserFlag.VERIFIED_DEVELOPER -> DISCORD_VERIFIED_DEVELOPER
            else -> null
        }
    }

    /*private fun OnlineStatus.toEmote() = when (this) {
        OnlineStatus.ONLINE -> "<:online2:464520569975603200>"
        OnlineStatus.IDLE -> "<:away2:464520569862357002>"
        OnlineStatus.DO_NOT_DISTURB -> "<:dnd2:464520569560498197>"

        else -> "<:offline2:464520569929334784>"
    }*/

    private fun OffsetDateTime.toBoostEmote(): String {
        return when (this.until(OffsetDateTime.now(), ChronoUnit.MONTHS)) {
            0L, 1L -> {
                "<:booster:738374009300975686>"
            }
            2L -> {
                "<:booster2:738374044247654480>"
            }
            3L -> {
                "<:booster3:738374173159850054>"
            }
            else -> {
                "<:booster4:738374213970165782>"
            }
        }
    }
}
